### ISSUE TITLE
Automatic refresh of access token

### ISSUE THREAD

**Issue by henk23** [jwilsson/spotify-web-api-php#137]
Hey, if I did not miss anything we have two options to notice that an access token refresh is needed:

- An api call throws a SpotifyWebAPIException with message 'access token expired'.
  - Then we have to refresh the token and run the call again.
- We check the expirationTime before the request.
  - Then we have to refresh the token and run the call.

To me this seems something that could be automated, because basically every developer using this library is going to need this at some point.
I thought of something like this:

    class SpotifyWebAPI {
        public function __construct($request = null, $autoRefreshAccessToken = true) {
        // ...

I could try to make a PR if this would be desired functionality.

**Comment by jwilsson** [comment]
Hi!
Thanks for the suggestion!

I don't know how feasible this would be with the current architecture, given how separated the `Session` and `SpotifyWebAPI` classes are. They don't know anything about each other and since the user of the library is responsible for passing access token between them I don't know how this could be automated I'm afraid. 

You're obviously more than welcome to try something out but I can't promise it'll be included in the library. My only suggestion if you do try to do something would be to check the expiration time before each call and refresh the token as needed.

**Comment by henk23** [comment]
Hey. I tried and failed. ;)
I ended up writing a `SpotifyService` to wrap `Session` and `SpotifyWebAPI`, checking the expirationTime before returning a valid api instance. But since the storage of the accesToken, refreshToken and expirationTime depends on the particular project, I also don't see how to implement this in a general way to include in this repo.

**Comment by adamazad** [comment]
I wrote a code to swap token 10 minutes before they expire (PHP run via
Node). I have a project that let's you schedule adding tracks to playlist
and this required fresh tokens at any given time.

**Comment by SanderSander** [comment]
It should be possible to detect that the access token is expired, and on that moment use the refresh token to obtain a new one.

From the docs: 

> Access tokens are deliberately set to expire after a short time, after which new tokens may be granted by supplying the refresh token originally obtained during the authorization code exchange.

Ofcourse it would be more efficient to check the `expires_in`, instead for waiting for the expired status code. 

But when the `SpotifyWebAPI` has a method like `setSession(Session $session)` it should be possible to auto refresh the access token. Even when a expire response is returned by spotify.

**Comment by SanderSander** [comment]
Ok after some testing this worked for me, it isn't really nice YET. but it works without changing any code in the repository. 

This could be added without breaking anything, and with tests and fixing the ugly stuff (Like `setAPI`), so this feature is doable.  

```
class SessionBasedRequest extends Request
{
    private const TOKEN_EXPIRED = "The access token expired";

    /**
     * @var Session
     */
    private $session;

    /**
     * @var SpotifyWebAPI
     */
    private $api;

    public function __construct(Session $session)
    {
        $this->session = $session;
    }

    public function setAPI(SpotifyWebAPI $api)
    {
        $this->api = $api;
    }

    public function send($method, $url, $parameters = [], $headers = [])
    {
        try {
            return parent::send($method, $url, $parameters, $headers);
        }
        catch (SpotifyWebAPIException $exception)
        {
            if ($exception->getCode() === 401 && $exception->getMessage() === self::TOKEN_EXPIRED) {
                $this->session->refreshAccessToken($this->session->getRefreshToken());
                if($this->api !== null) {
                    $this->api->setAccessToken($this->session->getAccessToken());
                }
                // Re-attempt the request with our refreshed access token
                $headers['Authorization'] = 'Bearer ' . $this->session->getAccessToken();
                return parent::send($method, $url, $parameters, $headers);
            }
            else {
                throw $exception;
            }
        }
    }
}
```

And the bootstrap code from README

```
require 'vendor/autoload.php';

$session = new SpotifyWebAPI\Session(
    'CLIENT_ID',
    'CLIENT_SECRET',
    'REDIRECT_URI'
);

// When refresh token is loaded elsewhere
// $session->setRefreshToken('REFRESH_TOKEN');

$request = new SessionBasedRequest($session);
$api = new SpotifyWebAPI\SpotifyWebAPI($request);
$request->setAPI($api);

if (isset($_GET['code'])) {
    $session->requestAccessToken($_GET['code']);
    $api->setAccessToken($session->getAccessToken());

    print_r($api->me());
} else {
    $options = [
        'scope' => [
            'user-read-email',
        ],
    ];

    header('Location: ' . $session->getAuthorizeUrl($options));
    die();
}
```
What do you guys think?

**Comment by doekenorg** [comment]
I have implemented roughly the same solution in my app. 

I think it would be a great asset if we'd have a specific exception for this situation. I'd like the library to check against the exception message and throw a `SpotifyWebApiExpiredTokenException`. 

Still think that wrapping your requests within a try catch is a valid way of dealing. I just don't like the extra check on the message.

**Comment by jwilsson** [comment]
@doekenorg I think adding a `SpotifyWebApiExpiredTokenException` would open up the door for too many highly specific exceptions. To me, there's two "types" of exceptions, API ones (missing parameters, invalid track ID etc.) and Auth ones (invalid client, expired token etc.). I'm not that keen on adding anymore exceptions at this point.

**Comment by doekenorg** [comment]
@jwilsson fair enough!

But since this is an auth error related, are you willing to add helpers to the Auth Exception? Something along the lines of `hasInvalidCredentials()` or `hasExpiredToken()`? These could be easy checks against the returned message. I just feel like checking against those messages in production code is like using `$something === 'true'`. It just doesn't feel right 😉, and it could break down easy when spotify decides to change it's exceptions messages. One update of this package would save everybody 😄 .

**Comment by jwilsson** [comment]
@doekenorg Sure, sounds like it could be helpful! Wanna send a PR for it?

**Comment by doekenorg** [comment]
Sure. Will do

**Comment by LewisHolt** [comment]
Any update on this?

**Comment by bulgariamitko** [comment]
my simple solution:

`try {
    $api->me();
} catch (Exception $e) {
    if ($e->getMessage() == 'The access token expired') {
        // Fetch the refresh token from somewhere. A database for example.
        $session->refreshAccessToken($refreshToken);
        $accessToken = $session->getAccessToken();
        // [Save this accessToken variable to the DB or Session
        // Set our new access token on the API wrapper and continue to use the API as usual
        $api->setAccessToken($accessToken);
    }
    echo '<pre>';
    print_r($e->getMessage());
    echo '</pre>';
    echo '<pre>';
    print_r('Refresh Token');
    echo '</pre>';
}
`

**Comment by jwilsson** [comment]
Hi everyone!
So after playing around with some of @SanderSander's ideas in https://github.com/jwilsson/spotify-web-api-php/issues/137#issuecomment-415789512 I've reached a solution that I'm pretty satisfied with and which I hope will solve everyone's needs.

I've pushed it to it's own branch and created [a PR for it](https://github.com/jwilsson/spotify-web-api-php/pull/167) for anyone interested to try out (install it with `composer require jwilsson/spotify-web-api-php:dev-auto-refresh`). It's still missing tests and some more docs (only the basics to get you up and running are there atm).

But please try it out and let me know of any issues, thoughts, and/or ideas!

**Comment by jwilsson** [comment]
This has been released in `2.11.0`! 🎉 

Docs are available [here](https://github.com/jwilsson/spotify-web-api-php/blob/89afa2b5e1386beae80eaf53fe4eb37bed8b873d/docs/examples/automatically-refreshing-access-tokens.md).
